### PR TITLE
 Add configuration item to allow kubeadm join to use a dns name pointing to control plane

### LINKED ISSUE #411: join w/ master dns name writes ip to kubelet config
i'm using kubeadm in a corporate cloud environment with enforced node recycling, one of the challenges of using kubeadm is that join will always use the ip address of the master instead of a passed in dns name (master cert also with subject alt to include dns name). the join ends up writing out the master ip to the kubelet.conf which complicates rotation/restoration of the master. given the symmetric secret, and the ca hash work being done in
https://docs.google.com/document/d/1SP4P7LJWSA8vUXj27UvKdVEdhpo5Fp0QHNo4TXvLQbw/edit?ts=5971498a#heading=h.5bejulk96xxi

is there any reason why kubeadm couldn't support using the actual value passed (ip or dns per user choice) instead of forcing ip usage?

kubeadm version - 1.7.4

### PR DESCRIPTION
**Pull request by stevesloka**
This adds a new flag (`--apiserver-advertise-dns-address`) to kubeadm which is used in node kubelet.confg to point to API server allowing users to define a DNS entry instead of an IP address.

Fixes kubernetes/kubeadm#411

```release-note
Adds new flag `--apiserver-advertise-dns-address` which is used in node kubelet.confg to point to API server
```
// @timothysc @craigtracey

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/addons.go [hunk 1]
```diff
@@ -145,6 +145,7 @@ func getAddonsSubCommands() []*cobra.Command {
 
 		if properties.use == "all" || properties.use == "kube-proxy" {
 			cmd.Flags().StringVar(&cfg.API.AdvertiseAddress, "apiserver-advertise-address", cfg.API.AdvertiseAddress, `The IP address or DNS name the API server is accessible on`)
+			cmd.Flags().StringVar(&cfg.API.AdvertiseDNSAddress, "apiserver-advertise-dns-address", cfg.API.AdvertiseDNSAddress, `The DNS name the API server is accessible on`)
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/addons.go]: I think I've turned a blind eye to some of the flag proliferation, this seems awfully redundant.  

/cc @errordeveloper

**Comment by stevesloka** [comment]
Need a way to allow users to specify dns name for API server as well as it address in the event the API server host machine has multiple. 

There are the San certs, however multiple can be passed so which one would you choose.

**Comment by stevesloka** [comment]
I started there when I began thinking through this PR to use `--apiserver-cert-extra-sans`, but you can pass multiple sans, so which one should I use? We could have a convention that the first one passed is the one to use, but thought that might be more confusing.

File: cmd/kubeadm/test/util.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func SetupMasterConfigurationFile(t *testing.T, tmpdir string, cfg *kubeadmapi.M
 		kind: MasterConfiguration
 		certificatesDir: {{.CertificatesDir}}
 		api:
-		   advertiseAddress: {{.API.AdvertiseAddress}}
+		   advertiseAddress: {{.API.AdvertiseDNSAddress}}
```
Comment (craigtracey) [inline cmd/kubeadm/test/util.go]: Is this change intentional? `AdvertiseAddress` should not be the same as `AdvertiseDNSAddress`

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go [hunk 3]
```diff
@@ -71,8 +71,10 @@ type MasterConfiguration struct {
 
 // API struct contains elements of API server address.
 type API struct {
-	// AdvertiseAddress sets the address for the API server to advertise.
+	// AdvertiseAddress sets the IP address for the API server to advertise.
 	AdvertiseAddress string `json:"advertiseAddress"`
+	// AdvertiseDNSAddress sets the DNS address for the API server to advertise.
+	AdvertiseDNSAddress string `json:"advertiseDNSAddress"`
```
Comment (craigtracey) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: I find this naming a bit confusing as I don't think anything is being "advertised."  While the `kubelet` uses the terminology of `advertiseAddress` to specify which IP the server is listening on, this new argument is being used only to indicate where a node should register itself.  Perhaps a more descriptive argument would be `controlPlaneEndpoint`?

**Review by craigtracey — CHANGES_REQUESTED** [review]
I think this is a pretty important change, and this will really help us close this gap. Thanks, @stevesloka!

**Comment by timothysc** [comment]
per sig conversation, we should leverage the config object to allow for updates and try not to proliferate cmd line flags b/c anything we do will have to be maintained post GA.

File: cmd/kubeadm/test/util.go (hunk 2 above)
Comment (stevesloka) [reply cmd/kubeadm/test/util.go]: Nope that's a mistake, thanks for catching.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go (hunk 3 above)
Comment (stevesloka) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go]: I think controlPlane make sense, I'll move to that, take out the args, and do a verify it works via the config.

File: cmd/kubeadm/app/cmd/phases/addons.go (hunk 1 above)
Comment (kapilt) [reply cmd/kubeadm/app/cmd/phases/addons.go]: part of the issue is the previous advertiseaddress was actually a listen address not an advertise.

**Comment by stevesloka** [comment]
@timothysc @craigtracey I updated to only use the config, also rebased and pushed.

File: cmd/kubeadm/app/cmd/phases/controlplane.go [hunk 4]
```diff
@@ -143,7 +143,7 @@ func getControlPlaneSubCommands(outDir, defaultKubernetesVersion string) []*cobr
 		cmd.Flags().StringVar(&cfg.KubernetesVersion, "kubernetes-version", cfg.KubernetesVersion, `Choose a specific Kubernetes version for the control plane`)
 
 		if properties.use == "all" || properties.use == "apiserver" {
-			cmd.Flags().StringVar(&cfg.API.AdvertiseAddress, "apiserver-advertise-address", cfg.API.AdvertiseAddress, "The IP address or DNS name the API server is accessible on")
+			cmd.Flags().StringVar(&cfg.API.AdvertiseAddress, "apiserver-advertise-address", cfg.API.AdvertiseAddress, "The IP address of the API server is accessible on")
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: So I need to look through history here.  I think other changes that crib'd hostIP from self-hosting affected this.  Sorry it's taking a while but we want the original intent to be preserved.  @kad if you have a moment could you dig the change, otherwise i'll have to search for it tomorrow.

Comment (craigtracey) [reply cmd/kubeadm/app/cmd/phases/controlplane.go]: This is actually a defect. The `kube-apiserver` only accepts an IP for `--advertise-address`.
https://kubernetes.io/docs/reference/generated/kube-apiserver/

I believe what @stevesloka has here is correct.

**Comment by stevesloka** [comment]
Does anyone have time this week to review? Curious if the implementation is acceptable or if anyone has any additional comments.

**Comment by timothysc** [comment]
/approve 

@kubernetes/sig-cluster-lifecycle-pr-reviews - For transparency I'd prefer someone non-heptio to lgtm

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 5]
```diff
@@ -548,5 +548,10 @@ func getAltNames(cfg *kubeadmapi.MasterConfiguration) (*certutil.AltNames, error
 		}
 	}
 
+	// add api server dns advertise address
+	if len(cfg.API.ControlPlaneEndpoint) > 0 {
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/phases/certs/certs.go]: Shouldn't we also validate that `ControlPlaneEndpoint` is a proper DNS subdomain, i.e., `validation.IsDNS1123Subdomain(ControlPlaneEndpoint)`?

File: cmd/kubeadm/app/util/endpoint.go [hunk 6]
```diff
@@ -37,15 +38,21 @@ func GetMasterEndpoint(cfg *kubeadmapi.MasterConfiguration) (string, error) {
 // GetMasterHostPort returns a properly formatted Master IP/port pair or error
 // if the IP address can not be parsed or port is outside the valid TCP range.
 func GetMasterHostPort(cfg *kubeadmapi.MasterConfiguration) (string, error) {
-	masterIP := net.ParseIP(cfg.API.AdvertiseAddress)
-	if masterIP == nil {
-		return "", fmt.Errorf("error parsing address %s", cfg.API.AdvertiseAddress)
+	var masterIP string
+	if len(cfg.API.ControlPlaneEndpoint) > 0 {
+		masterIP = cfg.API.ControlPlaneEndpoint
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/util/endpoint.go]: `validation.IsDNS1123Subdomain(ControlPlaneEndpoint)` here as well, and if the validation fails, error out.

File: cmd/kubeadm/app/util/endpoint.go [hunk 7]
```diff
@@ -37,15 +38,21 @@ func GetMasterEndpoint(cfg *kubeadmapi.MasterConfiguration) (string, error) {
...
+	} else {
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/util/endpoint.go]: I'm also wondering what if when both `cfg.API.AdvertiseAddress` and `cfg.API.ControlPlaneEndpoint` are configured. The current change only returns `cfg.API.ControlPlaneEndpoint` as master IP. But this seems ok, anyway.

**Review by xiangpengzhao — COMMENTED** [review]
a couple of concerns. otherwise LGTM.

**Comment by errordeveloper** [comment]
@stevesloka I think you might wanna take a look at `TestPrintConfiguration` failure, seems like it could be indirectly related to these changes.

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 5 above)
Comment (stevesloka) [reply cmd/kubeadm/app/phases/certs/certs.go]: I added this and also covered with tests.

**Comment by stevesloka** [comment]
@errordeveloper I got those tests fixed up, looks like the build is good now.

File: cmd/kubeadm/app/util/endpoint.go [hunk 8]
```diff
@@ -37,15 +39,25 @@ func GetMasterEndpoint(cfg *kubeadmapi.MasterConfiguration) (string, error) {
...
 // if the IP address can not be parsed or port is outside the valid TCP range.
 func GetMasterHostPort(cfg *kubeadmapi.MasterConfiguration) (string, error) {
-	masterIP := net.ParseIP(cfg.API.AdvertiseAddress)
-	if masterIP == nil {
-		return "", fmt.Errorf("error parsing address %s", cfg.API.AdvertiseAddress)
+	var masterIP string
+	if len(cfg.API.ControlPlaneEndpoint) > 0 {
+		errs := validation.IsDNS1123Subdomain(cfg.API.ControlPlaneEndpoint)
+		if len(errs) > 0 {
+			return "", fmt.Errorf("error parsing `ControlPlaneEndpoint` to valid dns subdomain")
```
Comment (kad) [inline cmd/kubeadm/app/util/endpoint.go]: Maybe here we want to provide to user more information about error ? e.g. what is wrong with domain provided.

**Review by kad — COMMENTED** [review]
small nit: may I suggest to split autogenerated changes to separate commit ?

**Comment by stevesloka** [comment]
@kad I spit the commits into 2, one for code and the other for autogen bazel deps. Also, I have the validation of the DNS address return the errors from validation method, and updated the error message to show those errors which get returned.

**Comment by kad** [comment]
@stevesloka Thanks. Changes in `cmd/kubeadm/app/apis/kubeadm/v1alpha1/zz_generated.conversion.go` seems to be still in first commit. The rest looks good for me.

**Comment by stevesloka** [comment]
@kad I've moved the zz_generated over to the 2nd commit

**Comment by stevesloka** [comment]
Does anyone have any additional comments on this PR?

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by Amrutayan** [comment]
Same here? How to use this flag? @stevesloka 
unknown flag: --apiserver-advertise-dns-address
To see the stack trace of this error execute with --v=5 or higher

**Comment by Amrutayan** [comment]
Or are we allowing to pass DNS name under this flag? "--apiserver-advertise-address"

**Comment by Amrutayan** [comment]
Can someone update please?

**Comment by neolit123** [comment]
you should use --control-plane-endpoint / ControlPlaneEndpoint (in config) which is a kubeadm construct / concept. it's the endpoint that sits in front of API servers.
the api server advertise "endpoint" does not allow using DNS names so only IPs are used.

> controlPlaneEndpoint sets a stable IP address or DNS name for the control plane. It can be a valid IP address or a RFC-1123 DNS subdomain, both with optional TCP port. In case the controlPlaneEndpoint is not specified, the advertiseAddress + bindPort are used; in case the controlPlaneEndpoint is specified but without a TCP port, the bindPort is used. Possible usages are:

> In a cluster with more than one control plane instances, this field should be assigned the address of the external load balancer in front of the control plane instances.
In environments with enforced node recycling, the controlPlaneEndpoint could be used for assigning a stable DNS to the control plane.

https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/

for more questions please use #kubeadm on k8s slack.
